src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD201CancelAfterSwitchToMainThreadAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+                if (jtfSymbol != null)
+                {
+                    var switchMethods = jtfSymbol.GetMembers(Types.JoinableTaskFactory.SwitchToMainThreadAsync);
+                    var cancellationTokenTypeSymbol = startCtxt.Compilation.GetTypeByMetadataName(typeof(CancellationToken).FullName);
+                    startCtxt.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, switchMethods, cancellationTokenTypeSymbol)), SyntaxKind.InvocationExpression);
+                }
+            });
+        }
+
+        internal static ExpressionSyntax GetCancellationTokenInInvocation(InvocationExpressionSyntax invocation, SemanticModel semanticModel, INamedTypeSymbol cancellationTokenTypeSymbol)
+        {
+            // Consider that named arguments allow for alternative ordering.
+            return invocation.ArgumentList.Arguments.FirstOrDefault(arg => semanticModel.GetTypeInfo(arg.Expression).Type?.Equals(cancellationTokenTypeSymbol) ?? false)?.Expression;
+        }
+
+        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context, ImmutableArray<ISymbol> switchMethods, INamedTypeSymbol cancellationTokenTypeSymbol)
+        {
+            var invocationSyntax = (InvocationExpressionSyntax)context.Node;
+            var invokeMethod = context.SemanticModel.GetSymbolInfo(context.Node).Symbol as IMethodSymbol;
+            if (invokeMethod != null && switchMethods.Contains(invokeMethod))
+            {
+                var tokenSyntax = GetCancellationTokenInInvocation(invocationSyntax, context.SemanticModel, cancellationTokenTypeSymbol);
+                if (tokenSyntax == null)
+                {
+                    return;
+                }
+
+                // It this *actually* a CancellationToken?
+                if (!context.SemanticModel.GetTypeInfo(tokenSyntax).Type?.Equals(cancellationTokenTypeSymbol) ?? false)
+                {
+                    return;
+                }
+
+                var tokenSymbol = context.SemanticModel.GetSymbolInfo(tokenSyntax).Symbol;
+                if (tokenSymbol == null)
+                {
+                    return;
+                }
+
+                // Did a non-default CancellationToken get passed in?
+                if (tokenSymbol.Name == nameof(CancellationToken.None) &&
+                    tokenSymbol.ContainingType.Name == nameof(CancellationToken))
+                {
+                    return;
+                }
+
+                // Did the invocation get followed by a check on that CancellationToken?
+                var statement = invocationSyntax?.FirstAncestorOrSelf<AwaitExpressionSyntax>()?.FirstAncestorOrSelf<StatementSyntax>();
+                var containingBlock = statement?.Parent as BlockSyntax;
+                if (containingBlock != null)
+                {
+                    int statementIndex = containingBlock.Statements.IndexOf(statement);
+                    if (statementIndex >= 0)
```

Comment:
❔ Can this ever be false? It seems like this should always pass, and failure should result in some diagnostic being reported.

Reply:
I've removed it.